ase/sw/hssi/hssi_stream.c

```diff
@@ -0,0 +1,507 @@
+// Context for each chanel
+//
+typedef struct hssi_chan_context_s
+{
+#ifdef ASE_HSSI_EMUL_PCAP
+    // General PCAP handles
+    pcap_t *pcap_in;
+    pcap_t *pcap_out;
+    pcap_dumper_t *pcap_out_dump;
+    char errbuf[PCAP_ERRBUF_SIZE];
+    // Current packets
+    unsigned rx_len;
+    const u_char *rx_packet;
+    unsigned tx_len;
+    u_char tx_packet[65535];
+#else // not ASE_HSSI_EMUL_PCAP
+    // FIFOs with data
+    // FIFOs are filled via TX and read by RX
+    int tlast_fifo[LOOPBACK_FIFO_SIZE];
+    char *tdata_fifo[LOOPBACK_FIFO_SIZE];
+    char *tuser_fifo[LOOPBACK_FIFO_SIZE];
+    char *tkeep_fifo[LOOPBACK_FIFO_SIZE];
+    int sptr, eptr;
+#endif // ASE_HSSI_EMUL_PCAP
+} hssi_chan_context_t;
+
+hssi_chan_context_t chan_context[MAX_CHANNELS];
+
+#ifdef ASE_HSSI_EMUL_PCAP
+void _reset(int chan)
+{
+    // Close and clear everything
+    if (chan_context[chan].pcap_in != NULL)
+        pcap_close(chan_context[chan].pcap_in);
+    if (chan_context[chan].pcap_out != NULL)
+        pcap_close(chan_context[chan].pcap_out);
+    if (chan_context[chan].pcap_out_dump != NULL)
+        pcap_dump_close(chan_context[chan].pcap_out_dump);
+    memset((char *)&chan_context[chan], 0, sizeof(chan_context[chan]));
+
+    // Get the PCAPs dir
+    char pcaps_path[2000];
+    char *pcaps_path_envvar = "ASE_HSSI_PCAPS_PATH";
+    if(!getenv(pcaps_path_envvar))
+        hssi_error_and_kill("Environment variable %s not set.\n", pcaps_path_envvar);
+    sprintf(pcaps_path, "%s", getenv(pcaps_path_envvar));
+    // Open the pcaps
+    char str_buff[2048];
+    sprintf(str_buff, "%s/in%d.pcap", pcaps_path, chan);
+    chan_context[chan].pcap_in = pcap_open_offline(str_buff, chan_context[chan].errbuf);
+    sprintf(str_buff, "%s/out%d.pcap", pcaps_path, chan);
+    chan_context[chan].pcap_out = pcap_open_dead(DLT_EN10MB, 65535);
+    if (chan_context[chan].pcap_out == NULL)
+        hssi_error_and_kill("Failed to setup output PCAP - %s\n", str_buff);
+    chan_context[chan].pcap_out_dump = pcap_dump_open(chan_context[chan].pcap_out, str_buff);
+    if (chan_context[chan].pcap_out_dump == NULL)
+        hssi_error_and_kill("Failed to open output PCAP - %s\n", str_buff);
+}
+#else // not ASE_HSSI_EMUL_PCAP
+void _reset(int chan)
+{
+    for (int i=0; i<LOOPBACK_FIFO_SIZE; i++) {
+        chan_context[chan].tlast_fifo[i] = 0;
+        chan_context[chan].tdata_fifo[i] = (char *)ase_malloc(hssi_param_cfg.tdata_width_bits/8);
+        chan_context[chan].tuser_fifo[i] = (char *)ase_malloc(hssi_param_cfg.tuser_width_bits/8);
+        chan_context[chan].tkeep_fifo[i] = (char *)ase_malloc(hssi_param_cfg.tdata_width_bits/(8*8));
+    }
+    chan_context[chan].sptr = 0;
+    chan_context[chan].eptr = 0;
+}
+#endif // ASE_HSSI_EMUL_PCAP
+
+// ========================================================================
+//
+//  RX side functions
+//
+// ========================================================================
+#ifdef ASE_HSSI_EMUL_PCAP
+int set_next_rx(
```

Comment:
I confess I don't fully understand the PCAP emulation, but I'm surprised a bit by the complexity here. Could the ASE component of emulation be reduced to a very simple plugin interface: a function for sending a payload and a function for receiving payload? Ideally, we'd reduce core functions in the ASE library to very simple I/O and then have everything PCAP specific in your library.

ASE could provide as it's default plugin a loopback, also simple, but separate from the core shared send/receive functions.

Does that map to what you are building?

Reply:
Is the idea that the user will provide the plugin? So the user could actually provide any other plugin as well?

So basically ASE only defines the SV + communications part of the HSSI emulation and expects two functions to be defined by the plugin. It also provides the simple loopback plugin. Anything else is in the user's hand (or we can provide the user with PCAP plugin as well, but it is not native part of the ASE).

Also is there any other place where a plugin mechanism like this is already used in ASE (so I can look at it and fashion this one in a similar way)?

Reply:
The OPAE SDK uses a plugin architecture where dynamically loaded libraries are loaded at runtime. We could use a plugin architecture similar to that, picking loopback by default. I'm sure Tim or Ananda could point us to some simple templates for defining a couple functions and loading them from a plugin.

If that's too difficult, we could be less fancy and just link in the module that implements whatever emulator is needed. That case is simple: implement the same function names for HSSI read and write in a pair of modules, one for loopback and one for PCAP, and just link with one of them.

With either method, there ought to be some really simple core function that is the DPI-C entry point that then maps to whichever library/module is loaded. Everything common to managing the DPI-C data transfer can be in common code, leaving data semantics to the specific modules.